### DISCUSSION TITLE
Add support for TIME/TIME WITH TIME ZONE type

### DISCUSSION THREAD

**Discussion by wypb** [facebookincubator/velox#6792] [General]
Presto has TIME type: https://prestodb.io/docs/current/language/types.html#time, But currently Velox doesn't support it. To support the TIME type in Velox, I will do the following:

- Implement TIME as a logical type and use bigint to represent it.
- Since only Presto currently supports the TIME type, I will put it in `velox/functions/prestosql/types`, just like the TimestampWithTimeZone type.
- I will implement common cast operations of the TIME type, such as `cast(varchar as time)`, `cast(timestamp as time)`, `cast(time as varchar)`, just as implemented in [TimeOperators](https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/type/TimeOperators.java)
- Implementation of TIME type comparison operations, such as supporting `CAST(STR AS TIME) < TIME '13:01:11.001'`
- Function implementation related to TIME type, such as CheckSum, localtime, max,min, etc.

Thoughts? Cc: @mbasmanova @aditi-pandit @xumingming

**Comment by mbasmanova** [comment]
@wypb Thank you for helping with this.

Before deciding to implement TIME type as Presto-specific type, can you do a survey of database engines to find out which ones have a TIME type  and whether that type is similar / different from Presto's.

Please, investigate how we would need to modify function signatures and fuzzers to accommodate the new type and include this in the design.

> Function implementation related to TIME type, such as CheckSum, localtime, max,min, etc

Please, try to come up with as complete list of functions as possible.

> comparison operations

Please, try to come up with as complete list as possible.

> cast operations

Do we only need to support CAST to and from strings? If so, please, include a detailed description of the semantics. 

Feel free to create and attach a Google doc with all these details to keep things organized.

Thanks.

**Comment by xumingming** [comment]
Seems a lot of work, I can help if you need some @wypb

**Reply by mbasmanova** [reply]
@xumingming Yes, I expect this to be a substantial effort. Thank you, @xumingming, for offering help.

**Reply by wypb** [reply]
@xumingming Thank you for offering help. After the design discussion is over, I will see how much specific work needs to be done, and then see if it should be divided.

**Comment by wypb** [comment]
HI @mbasmanova I wrote a simple design document. Can you help me take a look at it when you have time? https://docs.google.com/document/d/1RmMR7EtbGljNXeix51Ek_YebKAFC4doCOpkuQbV0758/edit

**Reply by mbasmanova** [reply]
@wypb Thank you for putting together a design document. Can you give me comment or edit access?

**Reply by mbasmanova** [reply]
I read the "Database and computing engines support time type" section of the document. It seems to me that most systems support TIME type and some support TIME WITH TIME ZONE type. Some systems support milliseconds and others nanoseconds precision. Most systems use 64-bit integer for in-memory representation of TIME values. Open question: what are the options of in-meory representation for TIME WITH TIME ZONE values? 

I suggest to re-write "Database and computing engines support time type" section to (1) provide the above summary; (2) include a table that highlights the differences between different systems. Perhaps, use rows for database systems and columns for capabilities.

I suggest the following columns: TIME (is TIME type supported), TIME WITH TIME ZONE (is TIME WITH TIME ZONE type supported), Precision (ms/ns), In-Memory Representation.

The format used to render TIME and TIME WITH TIME ZONE type values can be a separate column in this table.

**Reply by mbasmanova** [reply]
BTW, it seems to make sense to consider introducing both TIME and TIME WITH TIME ZONE types together as they are related. Let's make sure the design includes both.

**Reply by mbasmanova** [reply]
> Velox supports time type simple design
> constexpr Time() : timeUnit_(TimeUnit::MILLISECOND), times_(0) {}

I expect TIME and TIME WITH TIME ZONE types to be logical types backed by physical type BIGINT. Hence, flat vectors of these types will be FlatVector<int64_t>. Note that it is wasteful to store timeUnit_ for every time value. 

If we need to support different precisions (do we?), we may want to introduce different types: TIME_MS and TIME_NS.

**Reply by mbasmanova** [reply]
> In VectorFuzzer, we can refer to randTimestamp to generate Time type data.

This doesn't look right. Timestamp values are not Time values, are they?

**Reply by mbasmanova** [reply]
> 4. Functions that support time type

It would be helpful to break up these functions into groups: scalar functions and aggregate functions.

**Reply by mbasmanova** [reply]
> 6. Cast operations that support time type

It would be helpful to describe the behavior of cast to/from varchar in detail.

**Reply by wypb** [reply]
HI @mbasmanova, Thank you for your reply, I have given you the editing permissions for that document. And I modified the document based on the comments above, but currently there is only a TIME type design. I will add a TIME WITH TIME ZONE type design tomorrow.

**Reply by mbasmanova** [reply]
@wypb Thank you for iterating on the doc. I left some further comments. 

@aditi-pandit It would be great if you could take a pass as well.

@xumingming Maybe start reviewing the design doc and think about how the work can be divided and which pieces you might want to pick up.

**Reply by wypb** [reply]
@mbasmanova I have just modified it according to the comments, please help to review it again when you have time, thank you.

**Reply by mbasmanova** [reply]
@wypb Thanks, but it doesn't look like the document has been updated to describe the design for both TIME and TIME WITH TZ types.

**Reply by wypb** [reply]
@mbasmanova I have added TIME WITH TZ type design today.

**Reply by mbasmanova** [reply]
@wypb Thank you for iterating on the design document.

I guess the question is whether to add distinct types: TimeMillisType, TimeMicrosType, TimeNanosType or a single parametric type Time(precision). What use cases other than Presto do we have in mind? Spark doesn't support TIME types. DuckDB doesn't use Velox. Arrow and Substrait support TIME time with multiple precisions in Arrow and micros in Substrait.

Would it make sense to add TimeMillisType Velox type and TimeWithTimeZone Presto-specific type?

@majetideepak @aditi-pandit @assignUser What do you think? Any preferences?

**Reply by xumingming** [reply]
Seems the primary use case for Velox is Presto & Spark, how about just support single Time type with milliseconds precision, in the future if other engine want to support other precision, it can extend the TimeType and create a new customized type?

**Reply by mbasmanova** [reply]
I think it is reasonable to add a single TIME type with milliseconds precision, but, perhaps, we should not name it TimeType, rather TimeMillisType. This way we can add more types in the future with different precisions and maintain consistent naming: TimeMicrosType, TimeNanosType. Extending TimeType is not an option since TimeType has a specific semantic (milliseconds precision).

**Reply by wypb** [reply]
I agree with the above discussion, that is, the current TIME type that only implements millisecond precision, and then belongs to the velox type; TIME WITH TZ also achieves millisecond accuracy and is considered a Presto-specific type. That would make it much easier for us to implement.

**Reply by xumingming** [reply]
@wypb I suggest the next step of this project:

* List all the tasks(e.g. task1, task2 .. task10)
* Maybe some of the tasks are already done in your private velox fork, you can mark them.
* If there are still remaining tasks, others(e.g. me) can help.
* You might need to submit some base PR about TimeType, so after it is merged you and others can work concurrently.

**Reply by assignUser** [reply]
@mbasmanova I don't have any specific preferences that have a technical backing (aesthetically I prefer the parametric type^^). With the heterogeneous landscape of time types out there the important thing is that what ever implementation we use we should make sure that we can extend it in the future with not too much effort and ideally in a backwards compatible maner (though I think at this time that is not too important due to development velocity? )

The support of relevant file formats for the choosen types might also be relevant e.g. in arrow we recently switched to parquet 2.6 as default which supports nanos https://github.com/apache/arrow/issues/35746

**Reply by aditi-pandit** [reply]
@mbasmanova : I'm in support of adding separate types. DECIMAL use of parameter is very essential since precision and scale could be any number and values intermingled between them. For TIME, we have distinct types each with a fixed interpretation and specific logic. Overloading all these in a single class could lead to convoluted code I think.

Lets add TimeMillisType and Presto TimeWithTimeZone.

**Reply by wypb** [reply]
Hi @aditi-pandit @xumingming @mbasmanova @assignUser  Thanks for everyone's replies.

Based on the above discussion, I split this feature into the following tasks. Please help me see if you have any other suggestions. Thank you.

### Tasks to be done

- **Task 1 [2 days, 10-24 ~ 10-25]**:  Add TimeMillisType and TimeWithTimeZone type implementations, TimeMillisType will be implemented as a Velox type and TimeWithTimeZone will be implemented as a Presto-specific type.
- **Task 2 [3 days, 10-26 ~ 10-30]**: All functions that support the time type need to confirm whether a type signature is required and add relevant test cases.
- **Task 3 [3 days, 10-31 ~ 11-02]**: Fuzzers support generating TIME/TIME WITH TIME ZONE  type data and adding test cases.
- **Task 4 [2 days, 11-03 ~ 11-06]**: Implement time type arithmetic operations and comparison operations, and add test cases.
- **Task 5 [4 days, 11-07 ~ 11-11]**: Implement cast implementation of TIME type and add related test cases.
- **Task 6 [4 days, 11-13 ~ 11-16]**: Implement cast implementation of TIME WITH TIME ZONE type and add related test cases.
- **Task 7 [2 days, 11-17 ~ 11-20]**: PrestoSerializer implements TIME WITH TIME ZONE type data serialization and deserialization. 
- **Task 8 [2 days, 11-21 ~ 11-22]**: Adapt the TIME/TIME WITH TIME ZONE  type in Prestissimo and add test cases.

**NOTE**: All tasks depend on the completion of task 1, so I will submit this MR first, and after the MR is merged, other tasks will be completed separately. In addition, except for task 1, there are no dependencies between other tasks and can be developed at the same time.

**Reply by mbasmanova** [reply]
@wypb Thank you for creating a list of tasks. Looks good. Would you update the list to add estimated time to completion? Are you committed to see this effort through all the way to the end?

**Reply by wypb** [reply]
@mbasmanova I can follow up on this work. In addition, many tasks can be developed simultaneously, @xumingming should also be able to help with some tasks.

**Reply by mbasmanova** [reply]
@wypb Sounds good. Please, add time / effort estimates to all tasks and provide a tentative timeline for delivery. @xumingming Please, identify tasks you'll be working on and help estimate time / effort needed to complete these.

**Reply by mbasmanova** [reply]
@wypb 

> Can be completed in one day

Does this include time to write tests and get through a code review? Please, make sure to include these.

For readability, put the estimate in the front:

```
Task 1 [n days]: Add TimeMillisType and TimeWithTimeZone type implementations, TimeMillisType will be implemented as a Velox type and TimeWithTimeZone will be implemented as a Presto-specific type 
```

**Reply by mbasmanova** [reply]
@aditi-pandit Aditi, would you be available to do initial review for PRs related to these tasks?

**Reply by wypb** [reply]
> Does this include time to write tests and get through a code review? Please, make sure to include these.

The above time does not include the time of code review. I have added 1 day for each task to process code review and corresponding modifications.

**Reply by mbasmanova** [reply]
@wypb Do you think 1 day is sufficient to get a PR reviewed and approved? This estimate feels a bit low. 

Would you also add a tentative timeline for when each task will be delivered and provide an estimate of completion date for the whole project?

**Reply by wypb** [reply]
> Do you think 1 day is sufficient to get a PR reviewed and approved? This estimate feels a bit low.

1 day should be relatively short, how about reserving 2 days? 

> Would you also add a tentative timeline for when each task will be delivered and provide an estimate of completion date for the whole project?

Ok

**Reply by wypb** [reply]
@mbasmanova I have submitted the code for task 1 https://github.com/facebookincubator/velox/pull/7212, please help to review this, thank you.

**Reply by aditi-pandit** [reply]
@mbasmanova : Yes, I can do a first round of reviews.

**Reply by mbasmanova** [reply]
@wypb Somehow I cannot find the PR for task 1. Would you share a link?

**Reply by wypb** [reply]
@mbasmanova  The link to this PR can be found [here](https://github.com/facebookincubator/velox/pull/7212)
